### PR TITLE
Fork-specific consensus-types interfaces

### PR DESCRIPTION
**Pull request by kasey**
**What type of PR is this?**

Other

**What does this PR do? Why is it needed?**

This brings back the changes from https://github.com/prysmaticlabs/prysm/pull/13928 that got lost in the shuffle of various electra branches.

The goal of this PR is to deprecate the following patterns:
- monolithic block body interface across all forks. This model requires all non-phase0 attributes getters to return an error that must be checked.
- monolithic ExecutionData interface. In addition to the unsupported getter error checking issue, these types also require new getters to be backported to every previous fork.

In place of these patterns, we can use interface composition to add a new interface at each fork. The ExecutionData version for a particular fork can be accessed with a vanilla go type assertion. The BeaconBlockBody equivalent requires a little more wiring, because there is only one underlying concrete type which satisfies the interfaces of all forks. So we can add safe cast helpers to `consensus-types/interfaces/cast.go`.

### CODE REVIEW

**Review by terencechain — DISMISSED** [review]

**Review by james-prysm — APPROVED** [review]

**Review by nisdas — APPROVED** [review]

File: consensus-types/interfaces/beacon_block.go [hunk 1]
```diff
@@ -125,8 +132,12 @@ type ExecutionData interface {
 	WithdrawalsRoot() ([]byte, error)
 	ValueInWei() (math.Wei, error)
 	ValueInGwei() (uint64, error)
-	DepositReceipts() ([]*enginev1.DepositReceipt, error)
-	WithdrawalRequests() ([]*enginev1.ExecutionLayerWithdrawalRequest, error)
+}
+
+type ExecutionDataElectra interface {
```
Comment (rkapka) [inline consensus-types/interfaces/beacon_block.go:137]: Shouldn't there be `ExecutionDataCapella` and `ExecutionDataDeneb` too? That will allow us to remove unnecessary methods from previous forks.

File: consensus-types/interfaces/cast.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
+package interfaces
+
+import "github.com/prysmaticlabs/prysm/v5/runtime/version"
+
+// AsROBlockBodyElectra safely asserts the ReadOnlyBeaconBlockBody to a ROBlockBodyElectra.
+// This allows the caller to access methods on the block body which are only available on values after
+// the Electra hard fork. If the value is for an earlier fork (based on comparing its Version() to the electra version)
+// an error will be returned. Callers that want to conditionally process electra data can check for this condition
+// and safely ignore it like `if err != nil && errors.Is(interfaces.ErrInvalidCast) {`
+func AsROBlockBodyElectra(in ReadOnlyBeaconBlockBody) (ROBlockBodyElectra, error) {
```
Comment (rkapka) [inline consensus-types/interfaces/cast.go:10]: Same as with execution data, shouldn't we have equivalent functions for all other forks?

File: consensus-types/interfaces/beacon_block.go (hunk 1 above)
Comment (kasey) [reply consensus-types/interfaces/beacon_block.go:137]: This is an incremental step, but yes I would like us to add interfaces for each fork.

File: consensus-types/interfaces/cast.go (hunk 2 above)
Comment (kasey) [reply consensus-types/interfaces/cast.go:10]: Yes, once we backport to the other forks.

File: consensus-types/blocks/execution.go [hunk 3]
```diff
@@ -980,23 +945,14 @@ func IsEmptyExecutionData(data interfaces.ExecutionData) (bool, error) {
...
 
-	drs, err := data.DepositReceipts()
-	switch {
-	case errors.Is(err, consensus_types.ErrUnsupportedField):
-	case err != nil:
-		return false, err
-	default:
+	epe, postElectra := data.(interfaces.ExecutionDataElectra)
+	if postElectra {
+		drs := epe.DepositReceipts()
```
Comment (potuz) [inline consensus-types/blocks/execution.go:950]: This pattern assumes that every `interfaces.ExecutionData` in future forks, that also satisfies `ExecutionDataElectra` will have to be handled in the same way. Say for example that `interfaces.ExecutionDataEpbs` satisfies the Electra interface, it requires different handling of deposits but it requires the same handling of 7002 withdrawals, etc. If we explicitly change it to not satisfy `ExecutionDataElectra` and satisfy a different `ExectutionDataEpbs` then we would have to copy paste the withdrawal handling. If we keep it satisfying `ExecutionDataElectra` then we would have to deal with deposits with an if-else `data.(interfaces.ExecutionDataEpbs)` within this block right?

Comment (kasey) [reply consensus-types/blocks/execution.go:950]: Copy/paste from team chat:

If the ssz is the same, then it's fine for `ExecutionDataEpbs` to satisfy `ExecutionDataElectra`. Satisfying the interface only dictates what kind of value are available, it does not dictate how the data should be processed. That should always be based on the slot (or version, which is based on the slot).

SSZ changes in nested types is similar to what we just went through with `Attestations`. In that case our solution was to move attestations to an interface. A benefit of this kind of composition is that we would only need to implement the logic to deal with the interface in 2 places.If we do want to actually deprecate (remove) a field, then yes one option would be to break the chain of interface composition. Another would be to backport an error to that fields signature. Another would be to allow the getter to return a nil value and rely on the caller to be fork-aware. Another option is to rely less on these big interfaces and move processing functions to accept narrower interfaces. For example, epbs could take:
```
type EpbsWithdrawalGetter interface {
    GetWithdrawals() *eth.PBSWithdrawal
}
```
whereas other methods would take:
```
type WithdrawalsGetter interface {
    GetWithdrawals() []byte
}
```
and then perform assertions on those instead of the whole big type. A problem in both cases is that now if we want block to have something like an `ExecutionData` type that the container could use to abstractly carry the payload. This type would have to become /increasingly/ abstract (some kind of `ConsensusType` interface that would have ssz methods etc, which is asserted downstream to narrower interfaces). Otherwise we would also break the chain of composition in the container (block body).

*TL;DR* this approach handles all the previous fork changes pretty well, with attestations being the only notable exception. If we see more extreme changes across forks I think the best way to deal with that is probably more abstract interfaces for nested types, and narrow interfaces used by consumers of those fields. We'll need to keep iterating on it to see what feels most maintainable in practice.
